emily/cdk/lib/emily-stack.ts

```diff
@@ -0,0 +1,196 @@
+// Import the AWS CDK
+import * as cdk from 'aws-cdk-lib';
+import * as apig from 'aws-cdk-lib/aws-apigateway';
+import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
+import * as iam from 'aws-cdk-lib/aws-iam';
+import * as lambda from 'aws-cdk-lib/aws-lambda';
+import { Construct } from 'constructs';
+import { EmilyStackProps } from './emily-stack-props';
+import { EmilyStackUtils } from './emily-stack-utils';
+
+/**
+* @class CloudFormationStack
+* @classdesc Creates a stack with DynamoDB tables and a Lambda function.
+*/
+export class EmilyStack extends cdk.Stack {
+
+    /**
+    * @constructor
+    * @param {Construct} scope The AWS CDK construct scope.
+    * @param {string} id The stack ID.
+    * @param {EmilyStackProps} props The stack properties.
+    */
+    constructor(scope: Construct, id: string, props: EmilyStackProps) {
+        super(scope, id, props);
+        const depositTable: dynamodb.Table = this.createOrUpdateDepositTable(props);
+        const withdrawalTable: dynamodb.Table = this.createOrUpdateWithdrawalTable(props);
+        const chainstateTable: dynamodb.Table = this.createOrUpdateChainstateTable(props);
+        const operationLambda: lambda.Function = this.createOrUpdateOperationLambda(
+            depositTable,
+            withdrawalTable,
+            chainstateTable,
+            props
+        );
+        const emilyApi: apig.SpecRestApi = this.createOrUpdateApi(operationLambda, props);
+    }
+
+    createOrUpdateDepositTable(props: EmilyStackProps): dynamodb.Table {
+        const tableId: string = 'DepositTable';
+        const table: dynamodb.Table = new dynamodb.Table(this, tableId, {
+            tableName: EmilyStackUtils.getResourceName(tableId, props),
+            partitionKey: {
+                name: 'BitcoinTxid',
+                type: dynamodb.AttributeType.BINARY,
+            },
+            sortKey: {
+                name: 'BitcoinTxOutputIndex',
+                type: dynamodb.AttributeType.NUMBER,
+            }
+        });
+
+        const indexName: string = "DepositStatus";
+        table.addGlobalSecondaryIndex({
+            indexName: indexName,
+            partitionKey: {
+                name: 'OpStatus',
+                type:  dynamodb.AttributeType.NUMBER
+            },
+            sortKey: {
+                name: 'LastUpdateHeight',
+                type:  dynamodb.AttributeType.NUMBER
+            },
+            projectionType: dynamodb.ProjectionType.INCLUDE,
+            nonKeyAttributes: [
+                "BitcoinTxid",
+                "BitcoinTxOutputIndex",
+                "Recipient",
+                "Amount",
+            ]
+        });
+
+        // TODO: Add an additional GSI for querying by user; not required for MVP.
+        return table;
+    }
+
+    createOrUpdateWithdrawalTable(props: EmilyStackProps): dynamodb.Table {
+        // Create DynamoDB table to store the messages. Encrypted by default.
+        const tableId: string = 'WithdrawalTable';
```

Comment:
You know, my first thought was whether we could merge the deposit and withdrawal tables but I'm not sure that would be simpler. I'm pointing it out to see if you have any thoughts on the matter.

Reply:
Yeah simpler to have two tables. Feel free to disregard.

Reply:
I originally designed it to be one table but it lead to inefficient storage and queries so I changed it to be two. To query by type you'd either need to make multiple queries on the big table and remove the ones of the type you don't care about.

Reply:
Global secondary indices are a wrapper around **just duplicating the table**, so introducing one for each a deposit and a withdrawal leads to the same amount of storage being used for the GSIs as two separate tables would, plus the whole size of the original table.

The GSIs included here have a restricted set of attributes that they duplicate, so it's not the whole table.

Reply:
> I originally designed it to be one table but it lead to inefficient storage and queries so I changed it to be two. To query by type you'd either need to make multiple queries on the big table and remove the ones of the type you don't care about.

I think you forgot the *or* part of the second sentence.

Reply:
> you'd either need to make multiple queries on the big table and remove the ones of the type you don't care about **...or make a GSI on the table that lets you do the query in one shot but duplicates all the data, at which point you have two tables anyway.**